yazi-plugin/src/external/highlighter.rs

```diff
@@ -56,51 +59,235 @@ impl Highlighter {
+		let mut before = Vec::with_capacity(if *plain { 0 } else { skip });
+		let mut after = Vec::with_capacity(area.height as usize);
+
+		let mut lines_handled = 0;
+		let mut buf = vec![];
+		while reader.read_until(b'\n', &mut buf).await.is_ok() {
+			lines_handled += 1;
+			if !*plain && buf.len() > MAX_LINE_BYTES_TO_PLAINTEXT_FALLBACK {
+				*plain = true;
+				drop(mem::take(&mut before));
+			}
+			let must_break = Self::handle_single_line(
+				lines_handled,
+				skip,
+				*plain,
+				area.height as usize,
+				&mut buf,
+				&mut before,
+				&mut after,
+			);
+			if must_break {
+				break;
 			}
 			buf.clear();
 		}
 
-		if skip > 0 && i < skip + limit {
-			return Err(PeekError::Exceed(i.saturating_sub(limit)));
+		let no_more_scroll = lines_handled < skip + area.height as usize;
+		if skip > 0 && no_more_scroll {
+			return Err(PeekError::Exceed(lines_handled.saturating_sub(area.height as usize)));
 		}
 
-		if plain {
-			let indent = " ".repeat(PREVIEW.tab_size as usize);
-			Ok(Text::from(after.join("").replace('\t', &indent)))
-		} else {
-			Self::highlight_with(before, after, syntax.unwrap()).await
+		Ok((before, after))
+	}
+
+	fn handle_single_line(
+		lines_handled: usize,
+		skip: usize,
+		plain: bool,
+		limit: usize,
+		line: &mut Vec<u8>,
+		before: &mut Vec<String>,
+		after: &mut Vec<String>,
+	) -> bool {
+		if line.is_empty() || lines_handled > skip + limit {
+			return true;
+		}
+
+		if line.ends_with(b"\r\n") {
+			line.pop();
+			line.pop();
+			line.push(b'\n');
+		} else if !line.ends_with(b"\n") {
+			line.push(b'\n')
+		}
+
+		let text = String::from_utf8_lossy(line).into_owned();
+		if lines_handled > skip {
+			after.push(text);
+		} else if !plain {
+			before.push(text);
+		}
+		false
+	}
+
+	fn chunk_by_width(line: Vec<u8>, width: usize) -> Vec<Vec<u8>> {
+		if String::from_utf8_lossy(&line).width() <= width {
+			return vec![line];
+		}
+
+		let mut resulted_lines = vec![];
+		let mut buf_line = vec![];
+		let mut buf_char = vec![];
+		let mut last_break_char_idx = 0;
+		let mut last_break_idx = 0;
+		for (i, byte) in line.iter().enumerate() {
+			let byte = *byte;
+
+			buf_line.push(byte);
+			buf_char.push(byte);
+
+			if let Ok(char) = String::from_utf8(buf_char.clone()) {
+				if ",.; ".contains(&char) {
+					last_break_char_idx = i + 1
+				}
+
+				let buf_line_width = String::from_utf8_lossy(&buf_line).width();
+				if buf_line_width < width {
+					buf_char.clear();
+					continue;
+				}
+
+				if last_break_char_idx == 0 {
+					// no spaces in line, break right here
+					match buf_line_width.cmp(&width) {
+						std::cmp::Ordering::Equal => {
+							resulted_lines.push(buf_line.clone());
+							buf_line.clear();
+							last_break_idx = i + 1;
+						}
+						std::cmp::Ordering::Greater => {
+							let take_bytes = buf_line.len() - buf_char.len();
+							buf_line = buf_line[..take_bytes].to_vec();
+							resulted_lines.push(buf_line.clone());
```

Comment:
Please remove all `.clone()` and use `mem::take()` instead.

Reply:
Done